### PR TITLE
Add IP Adapter training script and update the docs with instructions

### PR DESCRIPTION
**Pull request by AMohamedAakhil**
# What does this PR do?
This PR adds the [original IP Adapter](https://github.com/tencent-ailab/IP-Adapter) training scripts, and also updates the documentation with instructions on how to use it. 

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by AMohamedAakhil** [comment]
#7194

File: docs/source/en/training/ip_adapter.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  --data_root_path="{image_path}" \
+  --mixed_precision="fp16" \
+  --resolution=512 \
+  --train_batch_size=8 \
+  --dataloader_num_workers=4 \
+  --learning_rate=1e-04 \
+  --weight_decay=0.01 \
+  --output_dir="{output_dir}" \
+  --save_steps=10000
+```
```
Comment (sayakpaul) [inline docs/source/en/training/ip_adapter.md:69]: Lets first provide a single-GPU example command and then proceed to multi-GPU.

File: docs/source/en/training/ip_adapter.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+#### Description:
+The provided inference code is used to load a trained model checkpoint and extract the components related to image projection and IP (Image Processing) adapter. These components are then saved into a binary file for later use in inference.
+
+#### Usage Example:
+```python
+import torch
+
+# Load the trained model checkpoint
+ckpt = "checkpoint-50000/pytorch_model.bin"
```
Comment (sayakpaul) [inline docs/source/en/training/ip_adapter.md:98]: Let's default to serializing in `.safetensors` please as it's a more secure file-format.

File: examples/ip_adapter/README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+# Save the components into a binary file
+torch.save({"image_proj": image_proj_sd, "ip_adapter": ip_sd}, "ip_adapter.bin")
+```
+
+#### Parameters:
+- `ckpt`: Path to the trained model checkpoint file.
+- `map_location="cpu"`: Specifies that the model should be loaded onto the CPU.
+- `image_proj_sd`: Dictionary to store the components related to image projection.
+- `ip_sd`: Dictionary to store the components related to the IP adapter.
+- `"unet"`, `"image_proj_model"`, `"adapter_modules"`: Prefixes indicating components of the model.
```
Comment (sayakpaul) [inline examples/ip_adapter/README.md:121]: But we are not showing actual inference here no? I think we are just showing a part of the process. How to run inference with "ip_adapter.bin" is missing.

File: examples/ip_adapter/tutorial_train_sdxl.py [hunk 4]
```diff
@@ -0,0 +1,483 @@
...
+        if name.startswith("mid_block"):
+            hidden_size = unet.config.block_out_channels[-1]
+        elif name.startswith("up_blocks"):
+            block_id = int(name[len("up_blocks.")])
+            hidden_size = list(reversed(unet.config.block_out_channels))[block_id]
+        elif name.startswith("down_blocks"):
+            block_id = int(name[len("down_blocks.")])
+            hidden_size = unet.config.block_out_channels[block_id]
+        if cross_attention_dim is None:
+            attn_procs[name] = AttnProcessor()
```
Comment (sayakpaul) [inline examples/ip_adapter/tutorial_train_sdxl.py:361]: We should use `AttnProcessor2_0()` when using PyTorch 2.0 as it's more memory and compute efficient.

File: examples/ip_adapter/tutorial_train_sdxl.py [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+        else:
+            layer_name = name.split(".processor")[0]
+            weights = {
+                "to_k_ip.weight": unet_sd[layer_name + ".to_k.weight"],
+                "to_v_ip.weight": unet_sd[layer_name + ".to_v.weight"],
+            }
+            attn_procs[name] = IPAttnProcessor(hidden_size=hidden_size, cross_attention_dim=cross_attention_dim, num_tokens=num_tokens)
```
Comment (sayakpaul) [inline examples/ip_adapter/tutorial_train_sdxl.py:368]: Same as above.

File: examples/ip_adapter/tutorial_train_sdxl.py [hunk 6]
```diff
@@ -0,0 +1,483 @@
...
+    global_step = 0
+    for epoch in range(0, args.num_train_epochs):
+        begin = time.perf_counter()
+        for step, batch in enumerate(train_dataloader):
+            load_data_time = time.perf_counter() - begin
+            with accelerator.accumulate(ip_adapter):
+                # Convert images to latent space
+                with torch.no_grad():
+                    # vae of sdxl should use fp32
+                    latents = vae.encode(batch["images"].to(accelerator.device, dtype=torch.float32)).latent_dist.sample()
```
Comment (sayakpaul) [inline examples/ip_adapter/tutorial_train_sdxl.py:412]:
```suggestion
                    latents = vae.encode(batch["images"].to(accelerator.device, dtype=vae.dtype)).latent_dist.sample()
```

**Review by sayakpaul — COMMENTED** [review]
Thanks! I think the structure of the training scripts deviate quite a bit from how our official training scripts are written. For now, let's maybe put these under `research_projects`?

**Comment by AMohamedAakhil** [comment]
Sure!

**Comment by AMohamedAakhil** [comment]
Sure! And I'll work on the suggested changes

File: examples/ip_adapter/tutorial_train_sdxl.py (hunk 5 above)
Comment (AMohamedAakhil) [reply examples/ip_adapter/tutorial_train_sdxl.py:368]: Also imported IPAttnProcessor2_0 as IPAttnProcessor

Comment (sayakpaul) [reply examples/ip_adapter/tutorial_train_sdxl.py:368]: Better would be to first check if it's using PyTorch 2. and then dynamically selecting it. 

An example:
https://github.com/huggingface/diffusers/blob/869bad3e52c86a33452858901090dd86ff4bebe1/src/diffusers/models/attention_processor.py#L214

**Comment by sayakpaul** [comment]
To start, maybe we could add this to `research_projects`?

**Comment by sayakpaul** [comment]
@AMohamedAakhil we would very much like to add this to `research_projects`. Can we revive this PR?

**Comment by AMohamedAakhil** [comment]
Sure

**Comment by sayakpaul** [comment]
Thanks! Please let us know once done!

**Pull request closed without merging**